### ISSUE TITLE
Poor codegen for static const float arrays

### ISSUE THREAD

**Issue by zeux** [Microsoft/DirectXShaderCompiler#1398]
Given a shader with static const float3 array declared in a function, DXC generates really suboptimal codegen:

- The array is copied to a local variable (reading from that variable may require spilling it from registers to local memory)
- The array is lazily initialized using a bool variable

```
float4 main(uint iid: SV_InstanceId): SV_Position
{
    static const float3 uvRoutingA[16] = { float3(0,0,1),  float3(1,0,0), float3(1,0,0), float3(0,0,1),  float3(1,0,0), float3(1,0,0),  float3(1,0,0),        float3(1,0,0),           float3(0,0,1),        float3(0,0,0), /* #10: */ float3(1,0,0), float3(0,1,0), float3(0,1,0), float3(1,0,0), float3(0,1,0),  float3(0,1,0) };

    return float4(uvRoutingA[iid], 1);
}
```

Here's the codegen for the code in question: http://shader-playground.timjones.io/37a1302a0c0b449c1abb324fbd005f0c
Ideally it seems like DXC should emit code that reads directly from the composite constant. This is important to make sure there's no opportunity for the driver to mess this up. I'm not sure what the interaction here is between frontend and backend optimizer like spirv-opt - the DXC output here is post-spirv-opt, with repeat spirv-opt runs producing no further optimizations.

**Comment by chaoticbob** [comment]
Thanks, Arseny. We're looking into it.

**Comment by zeux** [comment]
Just to clarify, this came up in a SPIRV-Cross -> MSL translation workflow; as it stands, the code produced results in bad performance on some drivers, presumably due to spilling. See http://shader-playground.timjones.io/f7857612d9a3064b611471dab89b986f

**Comment by antiagainst** [comment]
I see you've also filed a counterpart issue for Glslang: https://github.com/KhronosGroup/glslang/issues/1430

First regarding the boolean variable guarding the initialization. It is there because we need to make sure that function static variables are initialized once. Declaring a variable as `const` does not actually eliminate that need. For example, DXC/DXIL supports the following code:

```hlsl
float foo(float val) {
    static const float v = val;
    return v;
}

float4 PSMain() : SV_TARGET
{
  return foo(1) + foo(2);
}
```
(http://shader-playground.timjones.io/42fe6563ae020af82dd8f1e06214e882)

Its DXC/DXIL output:
```llvm
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float 2.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 1, float 2.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 2, float 2.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 3, float 2.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
```
Clearly we are writing to 2.0 to stage output variable here because of the `static`. Removing `static`, DXC/DXIL will write out 3.0. (Glslang is returning 3.0 even with `static`, which is a bug I think.)

We have no special handling for any function static variables. They are all guarded by boolean variables for initialization. We wanted to be "smarter" in the initial CodeGen stage to handle certain pattern and avoid the boolean guard, but as it turns out, the initializer can be very complicated, referring various function parameters, function local variables, function calls, etc. It's not clear where to cut the line as for what to specially handle and what not. So we are basically relying on spirv-opt to find suitable cases and optimize.

But for the case you pasted, spirv-opt is not optimizing the boolean guard away because we translate `uvRoutingA` and its boolean guard into `Private` storage class, and PrivateToLocalPass is not included in the current recipe of optimization passes. (PrivateToLocalPass is in legalization passes. But for this case, legalization is not needed. So DXC does not invoke legalization passes. And AFAICT, Glslang does not use the standard spirv-opt optimization recipe.)

So we will look into solving your second problem:

> The array is lazily initialized using a bool variable

hopefully by tweaking spirv-opt recipe.

As for the first problem,

> The array is copied to a local variable (reading from that variable may require spilling it from registers to local memory)

I'm not sure what is a good way to handle this. As you said in the Glslang issue, there is no way to index into a constant. One way I can think of is, maybe, we can turn the `OpStore` into `uvRoutingA` as a `OpVariable` inline initializer and hopefully the driver can do their work upon seeing a constant initilizer:

```llvm
%uvRoutingA = OpVariable %_ptr_Private__arr_v3float_uint_16 Private %const_array_value_instead_of_opstore
```

**Comment by antiagainst** [comment]
I've filed https://github.com/KhronosGroup/glslang/issues/1432 for the bug I mentioned in Glslang.

**Comment by johnkslang** [comment]
> They are all guarded by boolean variables for initialization. 

Is that necessary? Does HLSL prevent the transform of treating the static variable the same as a global-lifetime variable, with a global initializer, that happens to have lexical scoping for only one function?

That is, is the following not possible? Transform:

```
void f() { int v = expr; ... use of v ... }
```

to

```
__v = expr;
void f() { ... use of v... }
```

Does `expr` allow things only scoped within `f()` or otherwise not available at global scope?

**Comment by johnkslang** [comment]
Note this error from FXC when I tried to include a local, which implies the transform I want is possible:

`error X3691: invalid variable reference in static variable initializer.  Locals cannot be used to initialize static variables`

**Comment by zeux** [comment]
@antiagainst Yeah, unfortunately I keep discovering “SPIRV doesn’t work like this for unclear reasons” issues after I file these. If I were fixing this, I would omit the initializer guard for either transitively constexpr expressions or non-sideeffecting ones depending on the available analyses on frontend, but I would be content with spirv-opt handling this. As it stands, DXC output hurts performance on some Vulkan drivers as well - you can compare DXC and glslang on shader playground with Mali compiler.

Regarding indexing and spilling, the problem happens in context of MSL translation. It’s pretty catastrophic wrt performance but since there’s no way to clearly specify intent via SPIRV, fixing this, while crucial, would involve IR analysis on SPIRV-Cross side, sadly...

**Comment by antiagainst** [comment]
@johnkslang:

> Note this error from FXC when I tried to include a local, which implies the transform I want is possible:
> `error X3691: invalid variable reference in static variable initializer. Locals cannot be used to initialize static variables`

> Does `expr` allow things only scoped within `f()` or otherwise not available at global scope?

Yep. FXC does not support referencing function parameters/variables in the initializer of function static variables. However, DXC/DXIL supports it, even for much more complicated cases. For example,

```hlsl
struct S {
    int field;
    
    int getField() { return field; }
};

static S globalS = { 10 };

int foo(int input) {
    int local = 5.0;
    static const int val1 = local + input; // function local + function parameter
    static int val2[2] = {val1, val1 + globalS.getField()}; // method call
    return val2[1];
}

int VSMain(int input : A) : B {
    return foo(input) + foo(1000); // should return foo(input) * 2
}
```

(http://shader-playground.timjones.io/79c8af8f0f4dc60c4edcb9ce17d93b87)

DXC/DXIL correctly writes out `(input + 15) * 2`.

I'm not entirely sure whether it's the language evolution intention or merely a side-effect of switching to base on LLVM/Clang. But from my understanding, the long-term goal of HLSL is to be more C++-like. The flexibility here fits that. It's from the language's perspective, of course, and they are optimized away so no need for extra hardware capability.

> They are all guarded by boolean variables for initialization.
> Is that necessary?

We can do analysis on the initializer to tell whether it involves local symbols and promote the initialization into the entry function wrapper if not. That way we can avoid emitting the boolean guard. But then we need to analyze the initializer. It's doable, but still some work since the initializer can involve different kinds of AST nodes. On the other side, we already have all the mechanism in spirv-opt. Just need to register PrivateToLocalPass to the standard optimization pass. I believe that pass is generally good for optimization. So it seems to me cleaning up this with spirv-opt is more suitable.

**Comment by antiagainst** [comment]
With #1404, the boolean guard should be optimized away for this case.
